tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaLevel.scala

```diff
@@ -0,0 +1,92 @@
+package at.forsyte.apalache.tla.lir
+
+/**
+ * The level of a TLA+ expression, as explained in: Lamport. Specifying Systems, p. 322.
+ */
+sealed trait TlaLevel extends Ordered[TlaLevel] {
+
+  /**
+   * The level number as assigned in the book.
+   */
+  val level: Int
+
+  /**
+   * A textual representation of the level.
+   */
+  val name: String
+
+  /**
+   * Compare this level to another level.
+   *
+   * @param that the level to compare
+   * @return 0, if the levels are equal; a negative number if `this` is smaller than `that`, and a positive number
+   *         if `this` is larger than `that`.
+   */
+  override def compare(that: TlaLevel): Int = {
+    level - that.level
+  }
+
+  /**
+   * Join the levels by computing the smallest level that covers both of `this` and `that`.
+   *
+   * @param that the level to join with
+   * @return the minimal level j that satisfies: `j <= this` and `j <= that`.
+   */
+  def join(that: TlaLevel): TlaLevel = {
+    TlaLevel.fromInt(Math.max(level, that.level))
+  }
+
+  /**
+   * Join the level with a sequence of level
+   *
+   * @param otherLevels a sequence of levels
+   * @return the join of `this` and otherLevels
+   */
+  def join(otherLevels: Seq[TlaLevel]): TlaLevel = {
+    otherLevels.foldLeft(this) { case (l, r) => l.join(r) }
+  }
+}
+
+object TlaLevel {
+  private val INT_TO_LEVEL = Seq(TlaLevelConst, TlaLevelState, TlaLevelAction, TlaLevelTemporal)
+
+  def fromInt(level: Int): TlaLevel = {
+    if (level < 0 || level >= INT_TO_LEVEL.length) {
+      throw new IllegalArgumentException(s"Unexpected level of TlaValue: $level")
+    } else {
+      INT_TO_LEVEL(level)
+    }
+  }
```

Comment:
Since we already have the `Int` values inhe `Const` objects, why not just have `fromInt` give `level.level`? It seems like this duplicates the logic: we record the int once as a position in the sequence and another time explicitly in the `level` attribute.

Reply:
Good point. Fixed that.